functions/go/set-name-prefix/main.go

```diff
@@ -0,0 +1,207 @@
+		}
+		prefix, err := getPrefix(resourceList.FunctionConfig)
+		if err != nil {
+			return fmt.Errorf("failed to get data field from function config: %w", err)
+		}
+		err = setPrefix(prefix, resMap, tc, pluginHelpers, plugin)
+		if err != nil {
+			return fmt.Errorf("failed to add name prefix %s: %w",
+				prefix.Prefix, err)
+		}
+		// update name back reference
+		err = nameref.FixNameBackReference(resMap)
+		if err != nil {
+			return fmt.Errorf("failed to fix name back reference: %w", err)
+		}
+
+		// remove kustomize build annotations
+		resMap.RemoveBuildAnnotations()
+		resourceList.Items, err = resMap.ToRNodeSlice()
+		if err != nil {
+			return fmt.Errorf("failed to convert resource map to items: %w", err)
+		}
+		return nil
+	})
+
+	cmd.Long = usage()
+	if err := cmd.Execute(); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+func usage() string {
+	return `Add a name prefix to all resources.
+
+Configured using a ConfigMap with the following key in 'data':
+
+prefix: Name prefix to add to resources.
+
+Example:
+
+To add a name prefix 'dev-' to all resources:
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-config
+data:
+  prefix: dev-
+
+Each function call can only add one name prefix. This function
+will keep idempotent across different calls. If a name of a
+resource has already had the same prefix then it will not add
+the prefix again.
+
+You can use key 'fieldSpecs' to specify the resource selector you
+want to use. By default, the function will not only add or update the
+labels in 'metadata/name' but also a bunch of different places where
+have references to the names. These field specs are defined in
+https://github.com/kubernetes-sigs/kustomize/blob/master/api/konfig/builtinpluginconsts/namereference.go
+
+To support your own CRDs you will need to add more items to fieldSpecs list.
+Your own specs will be used with the default ones.
+
+Field spec has following fields:
+
+- group: Select the resources by API version group. Will select all groups
+	if omitted.
+- version: Select the resources by API version. Will select all versions
+	if omitted.
+- kind: Select the resources by resource kind. Will select all kinds
+	if omitted.
+- path: Specify the path to the field that the value will be updated. This field
+	is required.
+- create: If it's set to true, the field specified will be created if it doesn't
+	exist. Otherwise the function will only update the existing field.
+
+For more information about fieldSpecs, please see 
+https://kubectl.docs.kubernetes.io/guides/extending_kustomize/builtins/#arguments-3
+
+Example:
+
+To add a name 'dev-' to path 'data/selector/name' in MyOwnResource:
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: my-config
+data:
+  prefix: dev-
+  fieldSpecs:
+  - path: data/selector/name
+    kind: MyOwnKind
+`
+}
+
+func getDefaultConfig() (transformerConfig, error) {
+	defaultConfigString := builtinpluginconsts.GetDefaultFieldSpecsAsMap()["nameprefix"]
```

Comment:
It seems currently we only use kustomize's [default config for nameprefix](https://github.com/kubernetes-sigs/kustomize/blob/master/api/konfig/builtinpluginconsts/nameprefix.go#L7).
I was expecting this function also update the name reference. If not, IMO this function won't provide a lot of value.
@frankfarzan WDYT?

Reply:
The name references are updated and that's the answer for your previous comment. The name references are updated in `nameref/nameref.go` file.

Reply:
Got it.
This confused me a bit because this function is called `getDefaultConfig` and it's in `main.go`. I thought it's the only place for loading all config. But `nameref/nameref.go` appear to have a function `getDefaultConfig` too.
I'd suggest load all configs in a places for better readability.